File: services/custom-shell-api/app/routes/media.py
from __future__ import annotations

import math
import mimetypes
import re
from typing import Optional
from urllib.parse import quote

from fastapi import APIRouter, Depends, File, Form, Header, HTTPException, UploadFile, status
from fastapi.responses import StreamingResponse
from sqlalchemy import and_, func
from sqlalchemy.orm import Session

from app.db import get_db
from app.dependencies import get_current_user, require_app_origin
from app.media_storage import (
    R2StorageNotConfiguredError,
    delete_from_r2,
    get_from_r2,
    iter_storage_body,
    upload_to_r2,
)
from app.models import CustomShellMedia, CustomShellUser, utcnow, uuid_str
from app.schemas import MediaBulkDeleteIn, MediaBulkDeleteOut, MediaListOut, MediaOut, MediaUpdateIn

router = APIRouter(prefix="/api/v1/media", tags=["media"])

IMAGE_TYPES = {"image/jpeg", "image/jpg", "image/png", "image/gif", "image/webp"}
VIDEO_TYPES = {"video/mp4", "video/webm", "video/quicktime", "video/x-msvideo", "video/x-matroska"}
ALLOWED_TYPES = IMAGE_TYPES | VIDEO_TYPES
IMAGE_MAX_BYTES = 10 * 1024 * 1024
VIDEO_MAX_BYTES = 100 * 1024 * 1024
FILENAME_SAFE_CHARS = re.compile(r"[^a-zA-Z0-9.-]+")


@router.get("", response_model=MediaListOut)
def list_media(
    page: int = 1,
    page_size: int = 20,
    file_type: Optional[str] = None,
    current_user: CustomShellUser = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> MediaListOut:
    page = max(1, page)
    page_size = min(max(1, page_size), 100)

    query = db.query(CustomShellMedia).filter(CustomShellMedia.user_id == current_user.id)
    if file_type is not None:
        if file_type not in {"image", "video"}:
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Invalid media type")
        query = query.filter(CustomShellMedia.file_type == file_type)

    total = query.with_entities(func.count(CustomShellMedia.id)).scalar() or 0
    rows = (
        query.order_by(CustomShellMedia.created_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )

    return MediaListOut(
        media=[_serialize_media(row) for row in rows],
        total=total,
        page=page,
        page_size=page_size,
        total_pages=math.ceil(total / page_size) if total else 0,
    )


@router.post("", response_model=MediaOut, status_code=status.HTTP_201_CREATED)
async def upload_media(
    file: UploadFile = File(...),
    alt_text: Optional[str] = Form(default=None),
    current_user: CustomShellUser = Depends(get_current_user),
    _: None = Depends(require_app_origin),
    db: Session = Depends(get_db),
) -> MediaOut:
    mime_type = file.content_type or "application/octet-stream"
    if mime_type not in ALLOWED_TYPES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid file type. Only images (JPEG, PNG, GIF, WebP) and videos (MP4, WebM, MOV, AVI, MKV) are allowed.",
        )

    data = await file.read()
    if not data:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="File is empty")

    file_type = "image" if mime_type in IMAGE_TYPES else "video"
    max_size = IMAGE_MAX_BYTES if file_type == "image" else VIDEO_MAX_BYTES
    max_size_label = "10MB" if file_type == "image" else "100MB"
    if len(data) > max_size:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"File size too large. Maximum size is {max_size_label}.")

    original_name = _clean_original_name(file.filename)
    stored_filename = _stored_filename(original_name, mime_type)
    storage_path = f"{current_user.id}/{stored_filename}"

    try:
        upload_to_r2(storage_path, data, mime_type)
    except R2StorageNotConfiguredError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="R2 storage is not configured. Set the CUSTOM_SHELL_R2_* environment variables for custom-shell-api.",
        ) from exc
    except Exception as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail="Upload failed") from exc

    row = CustomShellMedia(
        user_id=current_user.id,
        filename=stored_filename,
        original_name=original_name,
        alt_text=_clean_alt_text(alt_text),
        file_size=len(data),
        mime_type=mime_type,
        file_type=file_type,
        storage_path=storage_path,
    )

    try:
        db.add(row)
        db.commit()
        db.refresh(row)
    except Exception:
        db.rollback()
        delete_from_r2(storage_path)
        raise

    return _serialize_media(row)


@router.put("/{media_id}", response_model=MediaOut)
def update_media(
    media_id: str,
    payload: MediaUpdateIn,
    current_user: CustomShellUser = Depends(get_current_user),
    _: None = Depends(require_app_origin),
    db: Session = Depends(get_db),
) -> MediaOut:
    row = _get_owned_media(db, current_user.id, media_id)
    row.alt_text = _clean_alt_text(payload.alt_text)
    row.updated_at = utcnow()
    db.commit()
    db.refresh(row)
    return _serialize_media(row)


@router.delete("/{media_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_media(
    media_id: str,
    current_user: CustomShellUser = Depends(get_current_user),
    _: None = Depends(require_app_origin),
    db: Session = Depends(get_db),
):
    row = _get_owned_media(db, current_user.id, media_id)
    delete_from_r2(row.storage_path)
    db.delete(row)
    db.commit()
    return None


@router.post("/bulk-delete", response_model=MediaBulkDeleteOut)
def bulk_delete_media(
    payload: MediaBulkDeleteIn,
    current_user: CustomShellUser = Depends(get_current_user),
    _: None = Depends(require_app_origin),
    db: Session = Depends(get_db),
) -> MediaBulkDeleteOut:
    unique_ids = list(dict.fromkeys(payload.ids))
    rows = (
        db.query(CustomShellMedia)
        .filter(and_(CustomShellMedia.user_id == current_user.id, CustomShellMedia.id.in_(unique_ids)))
        .all()
    )

    for row in rows:
        delete_from_r2(row.storage_path)
        db.delete(row)

    db.commit()
    return MediaBulkDeleteOut(deleted_count=len(rows))


@router.get("/{media_id}/file")
def stream_media_file(
    media_id: str,
    range_header: Optional[str] = Header(default=None, alias="Range"),
    current_user: CustomShellUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    row = _get_owned_media(db, current_user.id, media_id)

    try:
        r2_object = get_from_r2(row.storage_path, range_header)
    except R2StorageNotConfiguredError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="R2 storage is not configured. Set the CUSTOM_SHELL_R2_* environment variables for custom-shell-api.",
        ) from exc
    except Exception as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail="Failed to load media") from exc

    headers = {
        "Accept-Ranges": "bytes",
        "Cache-Control": "private, max-age=31536000, immutable",
    }
    content_length = r2_object.get("ContentLength")
    content_range = r2_object.get("ContentRange")
    if content_length is not None:
        headers["Content-Length"] = str(content_length)
    if content_range:
        headers["Content-Range"] = str(content_range)

    return StreamingResponse(
        iter_storage_body(r2_object["Body"]),
        status_code=status.HTTP_206_PARTIAL_CONTENT if range_header and content_range else status.HTTP_200_OK,
        media_type=r2_object.get("ContentType") or row.mime_type,
        headers=headers,
    )


def _get_owned_media(db: Session, user_id: str, media_id: str) -> CustomShellMedia:
    row = (
        db.query(CustomShellMedia)
        .filter(CustomShellMedia.id == media_id, CustomShellMedia.user_id == user_id)
        .first()
    )
    if row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Media not found")
    return row


def _serialize_media(row: CustomShellMedia) -> MediaOut:
    return MediaOut(
        id=row.id,
        filename=row.filename,
        original_name=row.original_name,
        alt_text=row.alt_text,
        file_size=row.file_size,
        mime_type=row.mime_type,
        file_type=row.file_type,
        url=f"/api/v1/media/{row.id}/file?name={quote(row.filename)}",
        created_at=row.created_at.isoformat(),
        updated_at=row.updated_at.isoformat(),
    )


def _clean_original_name(filename: Optional[str]) -> str:
    name = (filename or "media").replace("\\", "/").split("/")[-1].strip()
    return name[:255] or "media"


def _stored_filename(original_name: str, mime_type: str) -> str:
    base, dot, extension = original_name.rpartition(".")
    if not dot:
        base = original_name
        extension = (mimetypes.guess_extension(mime_type) or "").lstrip(".")

    clean_base = FILENAME_SAFE_CHARS.sub("-", base).strip(".-") or "media"
    clean_extension = FILENAME_SAFE_CHARS.sub("", extension).strip(".")
    suffix = f".{clean_extension}" if clean_extension else ""
    return f"{uuid_str()}_{clean_base}{suffix}"[:255]


def _clean_alt_text(value: Optional[str]) -> Optional[str]:
    cleaned = value.strip() if value else ""
    return cleaned[:500] or None

File: apps/ai-agents/src/server/media.ts
import sanitizeHtml from "sanitize-html"
import { and, asc, desc, eq, sql } from "drizzle-orm"

import { db } from "@/server/db"
import { getPublicMediaUrl } from "@/server/media-storage"
import { aiAgentsMedia, type AiAgentsMedia } from "@/server/schema"
import { uuid } from "@/server/security"

export const IMAGE_TYPES = new Set([
  "image/jpeg",
  "image/jpg",
  "image/png",
  "image/gif",
  "image/webp",
  "image/svg+xml",
])
export const VIDEO_TYPES = new Set([
  "video/mp4",
  "video/webm",
  "video/quicktime",
  "video/x-msvideo",
  "video/x-matroska",
])
export const ALLOWED_TYPES = new Set([...IMAGE_TYPES, ...VIDEO_TYPES])

const IMAGE_MAX_BYTES = 10 * 1024 * 1024
const VIDEO_MAX_BYTES = 100 * 1024 * 1024
const FILENAME_SAFE_CHARS = /[^a-zA-Z0-9.-]+/g

export type MediaFileType = "image" | "video"
export type MediaSortBy =
  | "created_at"
  | "original_name"
  | "file_size"
  | "file_type"
export type MediaSortDirection = "asc" | "desc"

export type MediaItem = {
  id: string
  filename: string
  original_name: string
  alt_text: string | null
  file_size: number
  mime_type: string
  file_type: MediaFileType
  url: string
  created_at: string
  updated_at: string
}

export type MediaListResponse = {
  media: MediaItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export function getMediaFileType(mimeType: string): MediaFileType {
  return IMAGE_TYPES.has(mimeType) ? "image" : "video"
}

export function validateMediaFile(mimeType: string, size: number) {
  if (!ALLOWED_TYPES.has(mimeType)) {
    throw new Error(
      "Invalid file type. Only images (JPEG, PNG, GIF, WebP, SVG) and videos (MP4, WebM, MOV, AVI, MKV) are allowed."
    )
  }

  const fileType = getMediaFileType(mimeType)
  const maxSize = fileType === "image" ? IMAGE_MAX_BYTES : VIDEO_MAX_BYTES
  const maxSizeLabel = fileType === "image" ? "10MB" : "100MB"
  if (size > maxSize) {
    throw new Error(`File size too large. Maximum size is ${maxSizeLabel}.`)
  }
}

export function validateMediaContent(mimeType: string, data: Uint8Array) {
  if (mimeType === "image/svg+xml") {
    sanitizeSvgContent(data)
    return
  }

  const valid =
    (mimeType === "image/jpeg" || mimeType === "image/jpg") &&
      hasPrefix(data, [0xff, 0xd8, 0xff]) ||
    mimeType === "image/png" &&
      hasPrefix(data, [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]) ||
    mimeType === "image/gif" &&
      (hasAscii(data, 0, "GIF87a") || hasAscii(data, 0, "GIF89a")) ||
    mimeType === "image/webp" &&
      hasAscii(data, 0, "RIFF") &&
      hasAscii(data, 8, "WEBP") ||
    (mimeType === "video/mp4" || mimeType === "video/quicktime") &&
      hasAscii(data, 4, "ftyp") ||
    mimeType === "video/webm" &&
      hasPrefix(data, [0x1a, 0x45, 0xdf, 0xa3]) ||
    mimeType === "video/x-msvideo" &&
      hasAscii(data, 0, "RIFF") &&
      hasAscii(data, 8, "AVI ") ||
    mimeType === "video/x-matroska" &&
      hasPrefix(data, [0x1a, 0x45, 0xdf, 0xa3])

  if (!valid) {
    throw new Error("File content does not match the selected media type.")
  }
}

export function prepareMediaContent(mimeType: string, data: Uint8Array) {
  if (mimeType === "image/svg+xml") {
    return sanitizeSvgContent(data)
  }

  validateMediaContent(mimeType, data)
  return data
}

const svgSanitizeOptions = {
  allowedTags: "svg g path rect circle ellipse line polyline polygon title desc".split(" "),
  allowedAttributes: {
    svg: "xmlns viewBox width height role aria-label aria-labelledby fill stroke".split(" "),
    "*": "d x y x1 x2 y1 y2 cx cy r rx ry points fill stroke stroke-width opacity transform".split(" "),
  },
  parser: {
    lowerCaseAttributeNames: false,
    lowerCaseTags: false,
  },
} satisfies sanitizeHtml.IOptions

function sanitizeSvgContent(data: Uint8Array) {
  let source = ""
  try {
    source = new TextDecoder("utf-8", { fatal: true }).decode(data)
  } catch {
    throw new Error("File content does not match the selected media type.")
  }

  const sanitized = sanitizeHtml(source, svgSanitizeOptions).trim()
  if (
    !/^<svg(?:\s|>)/i.test(sanitized) ||
    /(?:javascript:|data:|url\s*\()/i.test(sanitized)
  ) {
    throw new Error("File content does not match the selected media type.")
  }

  return new TextEncoder().encode(sanitized)
}

export function cleanOriginalName(filename?: string) {
  const name = (filename || "media").replace(/\\/g, "/").split("/").pop()?.trim()
  return (name || "media").slice(0, 255)
}

function hasPrefix(data: Uint8Array, prefix: number[]) {
  return prefix.every((byte, index) => data[index] === byte)
}

function hasAscii(data: Uint8Array, offset: number, value: string) {
  if (data.length < offset + value.length) return false
  return Array.from(value).every(
    (character, index) => data[offset + index] === character.charCodeAt(0)
  )
}

export function storedFilename(originalName: string, mimeType: string) {
  const extensionIndex = originalName.lastIndexOf(".")
  const base =
    extensionIndex > -1 ? originalName.slice(0, extensionIndex) : originalName
  const originalExtension =
    extensionIndex > -1 ? originalName.slice(extensionIndex + 1) : ""
  const extension = originalExtension || defaultExtensionForMimeType(mimeType)
  const cleanBase = base.replace(FILENAME_SAFE_CHARS, "-").replace(/^[.-]+|[.-]+$/g, "") || "media"
  const cleanExtension = extension.replace(FILENAME_SAFE_CHARS, "").replace(/^\.+|\.+$/g, "")
  const suffix = cleanExtension ? `.${cleanExtension}` : ""
  return `${uuid()}_${cleanBase}${suffix}`.slice(0, 255)
}

export function cleanAltText(value?: string | null) {
  const cleaned = value?.trim() || ""
  return cleaned ? cleaned.slice(0, 500) : null
}

export async function listOwnedMedia({
  userId,
  page,
  pageSize,
  fileType,
  mimeType,
  sortBy = "created_at",
  sortDirection = "desc",
}: {
  userId: string
  page: number
  pageSize: number
  fileType?: MediaFileType
  mimeType?: "image/svg+xml"
  sortBy?: MediaSortBy
  sortDirection?: MediaSortDirection
}): Promise<MediaListResponse> {
  const normalizedPage = Math.max(1, page)
  const normalizedPageSize = Math.min(Math.max(1, pageSize), 100)
  const ownerWhere = eq(aiAgentsMedia.userId, userId)
  const typeWhere = fileType ? eq(aiAgentsMedia.fileType, fileType) : null
  const mimeWhere = mimeType ? eq(aiAgentsMedia.mimeType, mimeType) : null
  const where =
    typeWhere && mimeWhere
      ? and(ownerWhere, typeWhere, mimeWhere)
      : typeWhere
        ? and(ownerWhere, typeWhere)
        : mimeWhere
          ? and(ownerWhere, mimeWhere)
          : ownerWhere

  const [totalRow] = await db
    .select({ count: sql<number>`count(*)::int` })
    .from(aiAgentsMedia)
    .where(where)
  const total = totalRow?.count ?? 0
  const rows = await db
    .select()
    .from(aiAgentsMedia)
    .where(where)
    .orderBy(getMediaOrderBy(sortBy, sortDirection))
    .offset((normalizedPage - 1) * normalizedPageSize)
    .limit(normalizedPageSize)

  return {
    media: rows.map(serializeMedia),
    total,
    page: normalizedPage,
    page_size: normalizedPageSize,
    total_pages: total ? Math.ceil(total / normalizedPageSize) : 0,
  }
}

function getMediaOrderBy(sortBy: MediaSortBy, sortDirection: MediaSortDirection) {
  const column =
    sortBy === "original_name"
      ? aiAgentsMedia.originalName
      : sortBy === "file_size"
        ? aiAgentsMedia.fileSize
        : sortBy === "file_type"
          ? aiAgentsMedia.fileType
          : aiAgentsMedia.createdAt

  return sortDirection === "asc" ? asc(column) : desc(column)
}

export async function getOwnedMedia(userId: string, mediaId: string) {
  const [row] = await db
    .select()
    .from(aiAgentsMedia)
    .where(and(eq(aiAgentsMedia.id, mediaId), eq(aiAgentsMedia.userId, userId)))
    .limit(1)

  if (!row) {
    throw new Error("Media not found")
  }

  return row
}

export function serializeMedia(row: AiAgentsMedia): MediaItem {
  return {
    id: row.id,
    filename: row.filename,
    original_name: row.originalName,
    alt_text: row.altText,
    file_size: row.fileSize,
    mime_type: row.mimeType,
    file_type: row.fileType as MediaFileType,
    url: getPublicMediaUrl(row.storagePath),
    created_at: row.createdAt.toISOString(),
    updated_at: row.updatedAt.toISOString(),
  }
}

function defaultExtensionForMimeType(mimeType: string) {
  if (mimeType === "image/jpeg" || mimeType === "image/jpg") return "jpg"
  if (mimeType === "image/png") return "png"
  if (mimeType === "image/gif") return "gif"
  if (mimeType === "image/webp") return "webp"
  if (mimeType === "image/svg+xml") return "svg"
  if (mimeType === "video/mp4") return "mp4"
  if (mimeType === "video/webm") return "webm"
  if (mimeType === "video/quicktime") return "mov"
  if (mimeType === "video/x-msvideo") return "avi"
  if (mimeType === "video/x-matroska") return "mkv"
  return ""
}
